src/app/InteractionModelEngine.cpp

```diff
@@ -408,6 +476,72 @@ Protocols::InteractionModel::Status InteractionModelEngine::OnReadInitialRequest
             });
         }
     }
+    else
+    {
+        System::PacketBufferTLVReader reader;
+
+        reader.Init(aPayload.Retain());
+
+        ReadRequestMessage::Parser readRequestParser;
+        VerifyOrReturnError(readRequestParser.Init(reader) == CHIP_NO_ERROR, Status::InvalidAction);
+
+        size_t requestedAttributePathCount = 0;
+        size_t requestedEventPathCount     = 0;
+
+        AttributePathIBs::Parser attributePathListParser;
+        bool hasValidAttributePath = false;
+
+        CHIP_ERROR err = readRequestParser.GetAttributeRequests(&attributePathListParser);
+        if (err == CHIP_NO_ERROR)
+        {
+            auto subjectDescriptor = apExchangeContext->GetSessionHandle()->AsSecureSession()->GetSubjectDescriptor();
+            err =
+                ParseAttributePaths(subjectDescriptor, attributePathListParser, hasValidAttributePath, requestedAttributePathCount);
+            if (err != CHIP_NO_ERROR)
+            {
+                return Status::InvalidAction;
+            }
+        }
+        else if (err != CHIP_ERROR_END_OF_TLV)
+        {
+            return Status::InvalidAction;
+        }
+
+        EventPathIBs::Parser eventpathListParser;
+        err = readRequestParser.GetEventRequests(&eventpathListParser);
+        if (err == CHIP_NO_ERROR)
+        {
+            TLV::TLVReader pathReader;
+            eventpathListParser.GetReader(&pathReader);
+            TLV::Utilities::Count(pathReader, requestedEventPathCount, false);
+        }
+        else if (err != CHIP_ERROR_END_OF_TLV)
+        {
+            return Status::InvalidAction;
+        }
+
+        if (requestedAttributePathCount == 0 && requestedEventPathCount == 0)
+        {
+            ChipLogError(InteractionModel,
+                         "Subscription from [%u:" ChipLogFormatX64 "] has zero path for attribute and event. Rejecting request.",
+                         apExchangeContext->GetSessionHandle()->GetFabricIndex(),
+                         ChipLogValueX64(apExchangeContext->GetSessionHandle()->AsSecureSession()->GetPeerNodeId()));
+            return Status::InvalidAction;
+        }
+
+        //
+        // TODO: We don't have an easy way to do a similar 'path expansion' for events to deduce
+        // access so for now, assume that the presence of any path in the event list means they
+        // might have some access to those events.
+        //
+        if (!hasValidAttributePath && requestedEventPathCount == 0)
+        {
+            ChipLogError(InteractionModel, "Read from [%u:" ChipLogFormatX64 "] has no access at all. Rejecting request.",
+                         apExchangeContext->GetSessionHandle()->GetFabricIndex(),
+                         ChipLogValueX64(apExchangeContext->GetSessionHandle()->AsSecureSession()->GetPeerNodeId()));
+            return Status::UnsupportedAccess;
```

Comment:
This does not match the spec.

Reply:
what should we return?

Reply:
Why are we returning an error at all?  It's not invalid to have empty things for a read, per spec, as far as I can tell.

Reply:
In fact, we should not be doing the pre-check for read at all, again as far as I can tell: that's not what the spec says to do.

Reply:
ok, let me remove pre-check for read.

Reply:
@andy31415 we don't need precheck for read per spec